src/module.rs

```diff
@@ -30,41 +30,96 @@ impl WasmModule {
         }
     }
 
+    pub fn imports(&self) -> impl Iterator<Item = &ImportEntry> {
+        self.module
+            .import_section()
+            .map_or(Either::Left(iter::empty()),
+                    |section| Either::Right(section.entries().iter()))
+    }
+
+    /// Safe function for counting the number of imported functions.
+    /// Use this instead of `self.imports().size_hint()`!
+    pub fn count_imported_functions(&self) -> usize {
+        self.module.import_count(ImportCountType::Function)
+    }
+
+    pub fn imported_functions(&self) -> impl Iterator<Item = WasmFunction> {
+        self.imports()
+            .enumerate()
+            .filter_map(move |(i, import)| if let External::Function(tyid) = import.external() {
+                     // NOTE: Unlike with Internal::Function(id),
+                     // the field of External::Function(_) is an index into
+                     // the type section.
+                     let name = import.field();
+                     let ty = self.get_type(*tyid)
+                         .expect(format!("Couldn't get type {} for imported function {}",
+                                         tyid,
+                                         name)
+                                         .as_str());
+
+                     Some(WasmFunction {
+                              // id is the index in the function index space.
+                              // An imported function's id is its order in the import section.
+                              id: i,
+                              ty,
+                              name: Some(name),
+                              body: None,
+                              source: SourceSection::Import,
+                          })
+                 } else {
+                     None
+                 })
+    }
+
+    /// Iterates over the function index space of the module.
+    /// According to the [WebAssembly design docs](https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md):
+    ///
+    /// > The function index space begins with an index for each imported
+    /// > function, in the order the imports appear in the Import Section,
+    /// > if present, followed by an index for each function in the Function Section,
+    /// > if present, in the order of that section.
     pub fn functions(&self) -> impl Iterator<Item = WasmFunction> {
         let function_count = self.module.functions_space();
         if function_count == 0 {
             return Either::Left(iter::empty::<WasmFunction>());
         }
 
-        let bodies = self.function_bodies();
-        let types = self.function_types();
-        assert_eq!(function_count, bodies.size_hint().0);
-        assert_eq!(function_count, types.size_hint().0);
+        let imported_functions = self.imported_functions();
```

Comment:
Do we use this here?

Reply:
Yes, it's chained with the functions section in the return value. I'll move the declaration down there.